include/oneapi/dpl/pstl/parallel_backend_utils.h

```diff
@@ -28,6 +28,72 @@ namespace dpl
 {
 namespace __utils
 {
+//------------------------------------------------------------------------
+// Uninitialized / initialized copy / move for trivial / non-trivial types
+//------------------------------------------------------------------------
+
+// struct __op_smart_ctor_assign - helper struct for uninitialized/initialized copy/move operations.
+// It can be used to perform copy/move operations for scalar and non-scalar types.
+template <typename _ValueType>
+struct __op_smart_ctor_assign
+{
+    using _ValueTypeDecayed = std::decay_t<_ValueType>;
+
+    // Uninitialized (for scalar types) / initialized (for other types) copy
+    /**
+     * @param _ValueTypeDecayed& __value_to - dest value
+     * @param const _ValueTypeDecayed& __value_from - source value
+     */
+    inline void
+    operator()(_ValueTypeDecayed& __value_to, const _ValueTypeDecayed& __value_from) const
+    {
+        if constexpr (std::is_scalar_v<_ValueTypeDecayed>)
+        {
+            __value_to = __value_from;
+        }
+        else
+        {
+            ::new (std::addressof(__value_to)) _ValueTypeDecayed(__value_from);
+        }
+    }
+
+    // Uninitialized (for scalar types) / initialized (for other types) move
+    /**
+     * @param _ValueTypeDecayed& __value_to - dest value
+     * @param _ValueTypeDecayed&& __value_from - source value
+     */
+    inline void
+    operator()(_ValueTypeDecayed& __value_to, _ValueTypeDecayed&& __value_from) const
+    {
+        if constexpr (std::is_scalar_v<_ValueTypeDecayed>)
+        {
+            __value_to = std::move(std::forward<_ValueTypeDecayed>(__value_from));
+        }
+        else
+        {
+            ::new (std::addressof(__value_to)) _ValueTypeDecayed(std::move(std::forward<_ValueTypeDecayed>(__value_from)));
+        }
+    }
+};
+
+// struct __op_smart_dtor - helper struct for destructor calls.
+// It can be used to perform destructor calls for types,
+// which are not trivially-destructible (we call destructor)
+// and for all other types (we doesn't call destructor).
+template <typename _ValueType>
+struct __op_smart_dtor
+{
+    static_assert(std::is_same_v<_ValueType, std::decay_t<_ValueType>>);
```

Comment:
Is there a reason why this helper requires the input type to be decayed but the assignment helper decays it internally?

Reply:
I think because (as in example at https://godbolt.org/z/vMT36djf3, see `sa1_ref`) for reference types we should avoid destructor calls. So I think we should require `decayed` types on caller side.